README.md
# @envoy/plugin-warn-if-update-available

warns if there is a newer version of CLI released

[![Version](https://img.shields.io/github/package-json/v/envoy/plugin-warn-if-update-available)](https://github.com/envoy/plugin-warn-if-update-available/packages/860241)
[![CircleCI](https://circleci.com/gh/envoy/plugin-warn-if-update-available/tree/master.svg?style=shield)](https://circleci.com/gh/envoy/plugin-warn-if-update-available/tree/master.svg?style=shield)
[![License](https://img.shields.io/github/license/envoy/plugin-warn-if-update-available)](https://github.com/envoy/plugin-warn-if-update-available/blob/master/package.json)

<!-- toc -->
* [@envoy/plugin-warn-if-update-available](#envoyplugin-warn-if-update-available)
* [What is this?](#what-is-this)
* [How it works](#how-it-works)
* [Installation](#installation)
* [Configuration](#configuration)
<!-- tocstop -->

# What is this?

This plugin shows a warning message if a user is running an out of date CLI.

![screenshot](./assets/screenshot.png)

# How it works

This checks the version against the npm registry asynchronously in a forked process, at most once per 7 days. It then saves a version file to the cache directory that will enable the warning. The upside of this method is that it won't block a user while they're using your CLI—the downside is that it will only display _after_ running a command that fetches the new version.

# Installation

Add the plugin to your project with `yarn add @envoy/plugin-warn-if-update-available`, then add it to the `package.json` of the oclif CLI:

```js
{
  "name": "mycli",
  "version": "0.0.0",
  // ...
  "oclif": {
    "plugins": ["@oclif/plugin-help", "@envoy/plugin-warn-if-update-available"]
  }
}
```

# Configuration

In `package.json`, set `oclif['warn-if-update-available']` to an object with
any of the following configuration properties:

- `timeoutInDays` - Duration between update checks. Defaults to 60.
- `nagTimeoutInDays` - Duration between upgrade nag prompts. Defaults to 0.
- `message` - Customize update message.

## Example configuration

```json
{
  "oclif": {
    "plugins": ["@envoy/plugin-warn-if-update-available"],
    "warn-if-update-available": {
      "timeoutInDays": 7,
      "nagTimeoutInDays": 1,
      "message": "<%= config.name %> update available from <%= chalk.greenBright(config.version) %> to <%= chalk.greenBright(latest) %>."
    }
  }
}
```


CHANGELOG.md
# [1.11.0](https://github.com/envoy/plugin-warn-if-update-available/compare/v1.10.0...v1.11.0) (2021-08-13)


### Features

* support nagTimeoutInDays to control how often to show the upgrade nag ([60cca0c](https://github.com/envoy/plugin-warn-if-update-available/commit/60cca0c074a32100460a3f7a609d55f3a6c59f0e))

# [1.10.0](https://github.com/envoy/plugin-warn-if-update-available/compare/v1.9.2...v1.10.0) (2021-08-13)


### Features

* move check-update to the postrun hook ([aa02e31](https://github.com/envoy/plugin-warn-if-update-available/commit/aa02e31d7232b523cfca220c80554148aff2c44c))

## [1.9.2](https://github.com/envoy/plugin-warn-if-update-available/compare/v1.9.1...v1.9.2) (2021-07-23)


### Bug Fixes

* update dependencies ([50df4f8](https://github.com/envoy/plugin-warn-if-update-available/commit/50df4f8fd56e16db3a9965cdb9a321620994453c))

## [1.9.1](https://github.com/envoy/plugin-warn-if-update-available/compare/v1.9.0...v1.9.1) (2021-06-24)


### Bug Fixes

* use outputJSON to write last-update-check file ([ca306e1](https://github.com/envoy/plugin-warn-if-update-available/commit/ca306e10c98cce255713bef7ba1757cd971489a2))

# [1.9.0](https://github.com/envoy/plugin-warn-if-update-available/compare/v1.8.5...v1.9.0) (2021-06-22)


### Bug Fixes

* correct imports for lodash.template and chalk ([b0d0ee6](https://github.com/envoy/plugin-warn-if-update-available/commit/b0d0ee677f2ae2fbb14668b831a7116556dce97e))


### Features

* readd chalk and lodash.template ([b0a34a7](https://github.com/envoy/plugin-warn-if-update-available/commit/b0a34a7b8fec69ac22fd19d25e6c011550177945))
* support timeOutInDays and message config ([db220a3](https://github.com/envoy/plugin-warn-if-update-available/commit/db220a3614a25d57471b75872c65f85258cc5db6))

## [1.8.5](https://github.com/envoy/plugin-warn-if-update-available/compare/v1.8.4...v1.8.5) (2021-06-22)


### Bug Fixes

* avoid constant checks when the host cli is unpublished ([9ec3c0f](https://github.com/envoy/plugin-warn-if-update-available/commit/9ec3c0f30bce52908107feccf085c6feb6a8589f))

## [1.8.4](https://github.com/envoy/plugin-warn-if-update-available/compare/v1.8.3...v1.8.4) (2021-06-22)


### Bug Fixes

* handle 404 from registry if the host cli hasn't published any packages ([800a546](https://github.com/envoy/plugin-warn-if-update-available/commit/800a54615af3d6ea96add0fcea90d93258d7e23c))

## [1.8.3](https://github.com/envoy/plugin-warn-if-update-available/compare/v1.8.2...v1.8.3) (2021-06-22)


### Bug Fixes

* prettier again ([17f6059](https://github.com/envoy/plugin-warn-if-update-available/commit/17f6059aaf11eceb22e5c6c43d6dab614b8be313))

## [1.8.2](https://github.com/envoy/plugin-warn-if-update-available/compare/v1.8.1...v1.8.2) (2021-06-22)


### Bug Fixes

* prettier ([8d0f4a4](https://github.com/envoy/plugin-warn-if-update-available/commit/8d0f4a466a04a9ceaa7c05161ad1c7e31af0ffb5))

## [1.8.1](https://github.com/envoy/plugin-warn-if-update-available/compare/v1.8.0...v1.8.1) (2021-06-22)


### Bug Fixes

* import libnpm correctly ([46def4a](https://github.com/envoy/plugin-warn-if-update-available/commit/46def4aed73e6670c17997e4f2bfd673b4ba86ef))

# [1.8.0](https://github.com/envoy/plugin-warn-if-update-available/compare/v1.7.0...v1.8.0) (2021-06-22)


### Bug Fixes

* configure the destination registry where the package will be published to ([200d0e5](https://github.com/envoy/plugin-warn-if-update-available/commit/200d0e553c57d2de99689fef93c750632f809c0a))
* lint issues ([903b6bd](https://github.com/envoy/plugin-warn-if-update-available/commit/903b6bd2ea826ca40aacfb7705dd010cba0eebb7))
* no explicit export ([c8eaf7b](https://github.com/envoy/plugin-warn-if-update-available/commit/c8eaf7b5181e57e580ad9997403a684ec029be08))
* remove org-global context ([386402d](https://github.com/envoy/plugin-warn-if-update-available/commit/386402db5de05fec7ab54ffbff7e001d84d323f4))
* silence warning about no type defs for libnpm ([eac4360](https://github.com/envoy/plugin-warn-if-update-available/commit/eac43600fdcfebb2ef64ac008ac2268e0c5ff995))
* update owner so we can publish to github private registry ([13b5eeb](https://github.com/envoy/plugin-warn-if-update-available/commit/13b5eeb9339cd0a8e15f80776bd096009c92390a))


### Features

* add eslint-config-prettier to silence formatting warnings ([37ccbd8](https://github.com/envoy/plugin-warn-if-update-available/commit/37ccbd80c76994f9197adf504bbf63e36331b96b))
* add required packages ([adb063a](https://github.com/envoy/plugin-warn-if-update-available/commit/adb063a67bf4ebd2ab41bf050ef9050e4f934564))
* remove unneeded packages ([61fa298](https://github.com/envoy/plugin-warn-if-update-available/commit/61fa298f79aa32234adbdee04adf36419150d700))
* replace implementation with one that respects npmrc and semver ([4e34773](https://github.com/envoy/plugin-warn-if-update-available/commit/4e347739f674e2d58e55e37293b496fe53ac06d6))

## [1.5.4](https://github.com/oclif/plugin-warn-if-update-available/compare/v1.5.3...v1.5.4) (2018-10-13)


### Bug Fixes

* remove greenkeeper badge ([f030bc1](https://github.com/oclif/plugin-warn-if-update-available/commit/f030bc1))

## [1.5.3](https://github.com/oclif/plugin-warn-if-update-available/compare/v1.5.2...v1.5.3) (2018-09-14)


### Bug Fixes

* remove greenkeeper-lockfile ([e95cd6c](https://github.com/oclif/plugin-warn-if-update-available/commit/e95cd6c))

## [1.5.2](https://github.com/oclif/plugin-warn-if-update-available/compare/v1.5.1...v1.5.2) (2018-08-28)


### Bug Fixes

* hide warning if updating ([6be6cd8](https://github.com/oclif/plugin-warn-if-update-available/commit/6be6cd8))

## [1.5.1](https://github.com/oclif/plugin-warn-if-update-available/compare/v1.5.0...v1.5.1) (2018-08-22)


### Bug Fixes

* rename oclif manifest to not be a dotfile ([84f870f](https://github.com/oclif/plugin-warn-if-update-available/commit/84f870f))

# [1.5.0](https://github.com/oclif/plugin-warn-if-update-available/compare/v1.4.0...v1.5.0) (2018-08-21)


### Features


src/hooks/postrun/check-update.ts
import { Hook } from "@oclif/config";
import * as libnpm from "libnpm";
import * as semver from "semver";
import * as fs from "fs-extra";
import * as path from "path";
import template from "lodash.template";
import chalk from "chalk";

const hook: Hook<"postrun"> = async function ({ config }) {
  const { name: packageName, version: currentVersion } = config;
  // Destructure package.json configuration with defaults
  const {
    timeoutInDays = 60,
    nagTimeoutInDays = 0,
    message = "<%= config.name %> update available from <%= chalk.greenBright(config.version) %> to <%= chalk.greenBright(latest) %>.",
  } = (config.pjson.oclif as any)["warn-if-update-available"] || {};
  const updateCheckPath = path.join(config.cacheDir, "last-update-check");
  const nagCheckPath = path.join(config.cacheDir, "last-nag-check");

  const isTimeoutStale = async (stalenessFile, timeout) => {
    try {
      const { mtime } = await fs.stat(stalenessFile);
      const staleAt = new Date(mtime.valueOf() + 1000 * 60 * 60 * 24 * timeout);
      return staleAt < new Date();
    } catch (error) {
      return true;
    }
  };

  const refreshNeeded = () => isTimeoutStale(updateCheckPath, timeoutInDays);

  const nagNeeded = () => isTimeoutStale(nagCheckPath, nagTimeoutInDays);

  // handle the situation where the host cli has not yet published a package. if this
  // plugin is installed on it, it will always fail the check since the registry will
  // return a 404. by catching the 404 and returning null, we get the chance to write the
  // last-update-check file so that we can avoid checking each time. the content being
  // null is a handled case in `checkVersion`.
  const readLatestRemoteManifest = async (): Promise<any | null> => {
    try {
      return await libnpm.manifest(
        `${packageName}@latest`,
        libnpm.config.read()
      );
    } catch (error) {
      if (error.code === "E404") {
        return null;
      }
      throw error;
    }
  };

  const checkForUpdate = async () => {
    const latestManifest = await readLatestRemoteManifest();

    await fs.outputJSON(updateCheckPath, latestManifest, {
      encoding: "utf8",
    });

    // eslint-disable-next-line no-use-before-define, @typescript-eslint/no-use-before-define
    await checkVersion();
  };

  const readLatestLocalManifest = async (): Promise<any | null> => {
    try {
      return await fs.readJSON(updateCheckPath, { encoding: "utf8" });
    } catch (error) {
      return null;
    }
  };

  const checkVersion = async () => {
    const latestManifest = await readLatestLocalManifest();

    // No version check has happened, so we can't tell if we're the latest version:
    if (latestManifest === null) {
      return null;
    }

    if (semver.lt(currentVersion, latestManifest.version)) {
      this.log(
        template(message)({
          chalk,
          config,
          latest: latestManifest.version,
        })
      );

      await fs.outputFile(nagCheckPath, "");
    }
  };

  if (await refreshNeeded()) {
    await checkForUpdate();
  } else if (await nagNeeded()) {
    await checkVersion();
  }
};

export default hook;
